addons/pkg-npm/jaxrs/src/main/java/org/commonjava/indy/pkg/npm/jaxrs/NPMContentAccessHandler.java

```diff
@@ -0,0 +1,424 @@
+                        // open the stream here to prevent deletion while waiting for the transfer back to the user to start...
+                        InputStream in = item.openInputStream( true, eventMetadata );
+                        final Response.ResponseBuilder builder = Response.ok( new TransferStreamingOutput( in ) );
+                        setInfoHeaders( builder, item, sk, path, true, contentController.getContentType( path ),
+                                        contentController.getHttpMetadata( item ) );
+                        if ( builderModifier != null )
+                        {
+                            builderModifier.accept( builder );
+                        }
+                        response = builder.build();
+                        // generating .http-metadata.json for npm group retrieve to resolve header requirements
+                        if ( eventMetadata.get( STORAGE_PATH ) != null && StoreType.group == st )
+                        {
+                            generateHttpMetadataHeaders( item, request, response );
+                        }
+                    }
+                }
+                finally
+                {
+                    if ( handleLocking )
+                    {
+                        item.unlock();
+                    }
+                }
+            }
+            catch ( final IOException | IndyWorkflowException e )
+            {
+                logger.error( String.format( "Failed to download artifact: %s from: %s. Reason: %s", path, name,
+                                             e.getMessage() ), e );
+                response = formatResponse( e, builderModifier );
+            }
+        }
+
+        logger.info( "RETURNING RESULT: {}:{}", sk, path );
+        return response;
+    }
+
+    private void generateNPMContentsFromTransfer( final Transfer transfer, final EventMetadata eventMetadata,
+                                                  Response response,
+                                                  final Consumer<Response.ResponseBuilder> builderModifier )
+    {
+        if ( transfer == null )
+        {
+            return;
+        }
+
+        OutputStream out1 = null;
+        OutputStream out2 = null;
+
+        Transfer target1;
+        Transfer target2;
+        String content1 = "";
+        String content2 = "";
+
+        ConcreteResource resource = transfer.getResource();
+        try
+        {
+            ObjectMapper mapper = new ObjectMapper();
+            JsonNode root = mapper.readTree( transfer.openInputStream( true ) );
+
+            String path1 = null;
+            String path2 = null;
+            JsonNode vnode = root.path( "versions" );
+            JsonNode anode = root.path( "_attachments" );
+            JsonNode idnode = root.path( "_id" );
+
+            if ( vnode.fields().hasNext() )
+            {
+                String version = vnode.fields().next().getKey();
```

Comment:
You might save the result of `vnode.fields().next()` then use it here for the key and below for the content. That way you can avoid an additional `findValue()` call...?

Reply:
Not fixed in new code?